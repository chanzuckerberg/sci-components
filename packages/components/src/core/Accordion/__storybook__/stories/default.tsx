import { Args } from "@storybook/react";
import RawAccordion, {
  AccordionDetails,
  AccordionHeader,
} from "src/core/Accordion";
import { EXTRA_SHORT_LOREM_IPSUM } from "src/common/storybook/loremIpsum";

export const Accordion = (props: Args): JSX.Element => {
  const { id, subtitle, useDivider, togglePosition } = props;
  return (
    <RawAccordion
      id={id}
      useDivider={useDivider}
      togglePosition={togglePosition}
      {...props}
    >
      <AccordionHeader id={`${id}-header`} subtitle={subtitle}>
        Accordion Header
      </AccordionHeader>
      <AccordionDetails>{EXTRA_SHORT_LOREM_IPSUM}</AccordionDetails>
    </RawAccordion>
  );
};
