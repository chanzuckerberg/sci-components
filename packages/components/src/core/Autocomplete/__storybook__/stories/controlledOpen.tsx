import { Args } from "@storybook/types";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { useState } from "react";
import RawAutocomplete from "src/core/Autocomplete";
import { AUTOCOMPLETE_WRAPPER_STYLES } from "../constants";

export const ControlledOpenDemo = <T extends DefaultAutocompleteOption>(
  props: Args
): JSX.Element => {
  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
  } = props;

  const [open, setOpen] = useState(true);

  return (
    <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
      <RawAutocomplete
        onClickAway={() => {
          setOpen(false);
        }}
        onClick={() => {
          if (open) {
            setOpen(false);
          } else {
            setOpen(true);
          }
        }}
        open={open}
        id="autocomplete-demo"
        disableCloseOnSelect={multiple}
        label={label}
        multiple={multiple}
        options={options}
        getOptionDisabled={(option: T) => {
          return option.name === "Type: feature request";
        }}
        {...props}
      />
    </div>
  );
};
