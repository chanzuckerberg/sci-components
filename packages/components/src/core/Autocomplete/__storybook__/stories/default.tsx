import { Args } from "@storybook/react";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { SyntheticEvent, useEffect, useState } from "react";
import RawAutocomplete, {
  AutocompleteMultiColumnValue,
  SDSAutocompleteValue,
} from "src/core/Autocomplete";
import TagFilter from "src/core/TagFilter";
import { AutocompleteValue } from "@mui/base";
import {
  AUTOCOMPLETE_DIV_MARGIN,
  AUTOCOMPLETE_WRAPPER_STYLES,
} from "../constants";

export const Autocomplete = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
>(
  props: Args
): JSX.Element => {
  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
    search,
    value: propValue,
    keepSearchOnSelect,
  } = props;

  type DisableClearable = false;
  type FreeSolo = false;

  const isControlled = propValue !== undefined;
  const [value, setValue] = useState<
    SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  >(
    (multiple ? [] : null) as SDSAutocompleteValue<
      T,
      Multiple,
      DisableClearable,
      FreeSolo
    >
  );

  const [selection, setSelection] = useState<string[]>([]);

  useEffect(() => {
    if (isControlled) {
      setValue(propValue);
    }
  }, [isControlled, propValue]);

  useEffect(() => {
    setSelection([]);
    setValue(
      (multiple ? [] : null) as SDSAutocompleteValue<
        T,
        Multiple,
        DisableClearable,
        FreeSolo
      >
    );
  }, [multiple, options]);

  return (
    <>
      <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
        <RawAutocomplete
          id="autocomplete-demo"
          disableCloseOnSelect={multiple}
          label={label}
          multiple={multiple}
          onChange={handleChange}
          keepSearchOnSelect={keepSearchOnSelect}
          options={options}
          search={search}
          value={value}
          getOptionDisabled={(option: T) => {
            return option.name === "Type: feature request";
          }}
          {...props}
        />
      </div>
      <div style={{ margin: AUTOCOMPLETE_DIV_MARGIN }}>
        {selection.length
          ? selection.map((item) => {
              return (
                <TagFilter
                  key={item}
                  label={item}
                  onDelete={() => handleTagDelete(item)}
                  onClick={() => handleTagDelete(item)}
                />
              );
            })
          : null}
      </div>
    </>
  );

  // eslint-disable-next-line sonarjs/cognitive-complexity
  function handleChange(
    _: SyntheticEvent<Element, Event>,
    newValue: SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  ) {
    setValue(newValue);
    const newSelection: string[] = [];

    if (
      (newValue as T)?.name ||
      (Array.isArray(newValue) && (newValue as T[])[0].name)
    ) {
      if (multiple) {
        setSelection(
          Array.isArray(newValue) ? newValue?.map((item) => item.name) : []
        );
      } else {
        if (newValue && !Array.isArray(newValue) && newValue.name) {
          setSelection([(newValue as T).name]);
        }
      }
    } else {
      if (multiple) {
        if (newValue) {
          Object.values(newValue).forEach((items) => {
            (items as T[])?.map(({ name }: { name: string }) => {
              newSelection.push(name);
            });
          });
          setSelection(newSelection);
        }
      } else {
        if (newValue) {
          Object.values(newValue).forEach((item) => {
            newSelection.push((item as T).name);
          });
        }
        setSelection(newSelection);
      }
    }
  }

  // eslint-disable-next-line sonarjs/cognitive-complexity
  function handleTagDeleteMultiColumn(tag: string) {
    if (multiple) {
      if (value) {
        Object.entries(value).forEach(([key, values]) => {
          const index = (values as T[])?.findIndex(
            (item: T) => item.name === tag
          );
          if (index > -1) {
            const newValue = [...(values as T[])];
            newValue.splice(index, 1);
            setValue((prev) => {
              return {
                ...prev,
                [key]: newValue as AutocompleteValue<
                  T,
                  Multiple,
                  DisableClearable,
                  FreeSolo
                >,
              };
            });

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    } else {
      if (value) {
        Object.entries(value).forEach(([key, val]) => {
          if ((val as T).name === tag) {
            const finalValue = { ...value } as AutocompleteMultiColumnValue<
              T,
              Multiple,
              DisableClearable,
              FreeSolo
            >;
            if (finalValue) delete finalValue[key];
            setValue(finalValue);

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    }
  }

  function handleTagDelete(tag: string) {
    if (
      (value && (value as T).name) ||
      (Array.isArray(value) && (value as T[])[0].name)
    ) {
      if (multiple && Array.isArray(value)) {
        const index = value?.findIndex((item) => item.name === tag);
        const newValue = [...value];
        newValue.splice(index, 1);
        setValue(
          newValue as SDSAutocompleteValue<
            T,
            Multiple,
            DisableClearable,
            FreeSolo
          >
        );

        const newSelection = [...selection];
        const deleteIndex = newSelection.indexOf(tag);
        newSelection.splice(deleteIndex, 1);
        setSelection(newSelection);
      } else {
        setValue(
          null as SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
        );
        setSelection([]);
      }
    } else {
      handleTagDeleteMultiColumn(tag);
    }
  }
};
