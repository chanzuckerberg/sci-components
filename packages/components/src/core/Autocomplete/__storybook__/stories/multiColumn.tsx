import { Args } from "@storybook/react";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_MULTI_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_MULTI_COLUMN_OPTIONS";
import { useEffect, useState } from "react";
import { AutocompleteMultiColumnValue } from "src/core/Autocomplete";
import {
  AutocompleteChangeDetails,
  AutocompleteChangeReason,
  AutocompleteValue,
} from "@mui/base";
import {
  AUTOCOMPLETE_DIV_MARGIN,
  AUTOCOMPLETE_WRAPPER_STYLES,
} from "../constants";
import TagFilter from "src/core/TagFilter";
import RawAutocompleteMultiColumn from "src/core/Autocomplete/components/AutocompleteMultiColumn";

export const AutocompleteMultiColumnDemo = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
>(
  props: Args
): JSX.Element => {
  const {
    multiple,
    options = AUTOCOMPLETE_MULTI_COLUMN_OPTIONS,
    search,
    label,
  } = props;

  type DisableClearable = false;
  type FreeSolo = false;

  const [selection, setSelection] = useState<string[]>([]);

  const [theValue, setTheValue] =
    useState<
      AutocompleteMultiColumnValue<T, Multiple, DisableClearable, FreeSolo>
    >();

  useEffect(() => {
    setSelection([]);
  }, [multiple]);

  const handleChange = (
    _event: React.SyntheticEvent,
    newValue: AutocompleteMultiColumnValue<
      T,
      Multiple,
      DisableClearable,
      FreeSolo
    >,
    _reason: AutocompleteChangeReason,
    _details?: AutocompleteChangeDetails<T>
  ) => {
    setTheValue(newValue);
    const newSelection: string[] = [];

    if (multiple) {
      if (newValue) {
        Object.values(newValue).forEach((items) => {
          (items as T[])?.map(({ name }: { name: string }) => {
            newSelection.push(name);
          });
        });
        setSelection(newSelection);
      }
    } else {
      if (newValue) {
        Object.values(newValue).forEach((item) => {
          newSelection.push((item as T).name);
        });
      }
      setSelection(newSelection);
    }
  };

  return (
    <>
      <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
        <RawAutocompleteMultiColumn<T, Multiple, DisableClearable, FreeSolo>
          label={label}
          disableCloseOnSelect={false}
          multiple={multiple}
          onChange={handleChange}
          options={options}
          search={search}
          value={theValue}
          {...props}
        />
      </div>
      <div style={{ margin: AUTOCOMPLETE_DIV_MARGIN }}>
        {selection.length
          ? selection.map((item) => {
              return (
                <TagFilter
                  key={item}
                  label={item}
                  onDelete={() => handleTagDelete(item)}
                  onClick={() => handleTagDelete(item)}
                />
              );
            })
          : null}
      </div>
    </>
  );

  function handleTagDelete(tag: string) {
    if (multiple) {
      if (theValue) {
        Object.entries(theValue).forEach(([key, values]) => {
          const index = (values as T[])?.findIndex(
            (item: T) => item.name === tag
          );
          if (index > -1) {
            const newValue = [...(values as T[])];
            newValue.splice(index, 1);
            setTheValue((prev) => {
              return {
                ...prev,
                [key]: newValue as AutocompleteValue<
                  T,
                  Multiple,
                  DisableClearable,
                  FreeSolo
                >,
              };
            });

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    } else {
      if (theValue) {
        Object.entries(theValue).forEach(([key, value]) => {
          if ((value as T).name === tag) {
            const finalValue = { ...theValue };
            delete finalValue[key];
            setTheValue(finalValue);

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    }
  }
};
