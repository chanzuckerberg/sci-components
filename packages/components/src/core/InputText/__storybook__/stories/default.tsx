import { Args } from "@storybook/react";
import RawInputText from "src/core/InputText";

export const InputText = (props: Args): JSX.Element => {
  const {
    id,
    intent,
    disabled,
    hideLabel,
    placeholder,
    label,
    sdsType,
    sdsStage,
  } = props;
  return (
    <RawInputText
      id={id}
      sdsType={sdsType}
      sdsStage={sdsStage}
      label={label}
      placeholder={placeholder}
      intent={intent}
      hideLabel={hideLabel}
      disabled={disabled}
      name="input-text-name"
      autoComplete="off"
    />
  );
};
