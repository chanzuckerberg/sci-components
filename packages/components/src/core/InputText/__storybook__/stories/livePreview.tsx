import { Args } from "@storybook/react";
import RawInputText from "src/core/InputText";
import { INPUT_TEXT_LIVE_PREVIEW_ROW_STYLES } from "../constants";

export const LivePreviewDemo = (props: Args): JSX.Element => {
  return (
    <div style={INPUT_TEXT_LIVE_PREVIEW_ROW_STYLES as React.CSSProperties}>
      <RawInputText
        id="textFieldPreview"
        label="Label"
        {...props}
        sdsType="textField"
        hideLabel
        placeholder="Value"
        style={{ width: "200px" }}
      />
      <RawInputText
        id="textAreaPreview"
        label="Label"
        {...props}
        sdsType="textArea"
        hideLabel
        placeholder="Value"
        style={{ minWidth: "200px" }}
        inputProps={{ style: { minWidth: "200px" } }}
      />
    </div>
  );
};
