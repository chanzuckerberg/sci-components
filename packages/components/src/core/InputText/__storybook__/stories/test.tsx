import { FormControl } from "@mui/material";
import { Args } from "@storybook/react";
import RawInputText from "src/core/InputText";

export const TestDemo = (props: Args): JSX.Element => {
  return (
    <FormControl>
      <RawInputText
        {...props}
        id="test-textField"
        sdsType="textField"
        label="Label"
        hideLabel={false}
        placeholder="Value"
        data-testid="inputTextBase"
      />

      {/* @ts-expect-error testing fail state */}
      <RawInputText
        sdsType="textField"
        hideLabel={false}
        data-testid="inputTextFail"
      />

      <RawInputText
        id="test-hide-label"
        sdsType="textField"
        label="Hidden Label"
        hideLabel
        data-testid="inputTextHideLabel"
      />

      <RawInputText
        id="test-textArea"
        sdsType="textArea"
        label="Label"
        hideLabel={false}
        placeholder="Value"
        data-testid="inputTextArea"
      />
    </FormControl>
  );
};
