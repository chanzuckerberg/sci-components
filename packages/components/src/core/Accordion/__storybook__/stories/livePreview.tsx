import { Args } from "@storybook/react";
import { Accordion } from "./default";

const livePreviewStyles = {
  display: "grid",
  gridColumnGap: "24px",
  gridRowGap: "50px",
  gridTemplateColumns: "repeat(4, 200px)",
};

export function LivePreviewDemo(props: Args): JSX.Element {
  return (
    <div style={livePreviewStyles as React.CSSProperties}>
      <Accordion
        id="accordion-1"
        togglePosition="right"
        useDivider
        {...props}
      />
      <Accordion
        id="accordion-2"
        togglePosition="right"
        subtitle="Optional Subtitle"
        useDivider
        {...props}
      />
      <Accordion id="accordion-3" togglePosition="left" useDivider {...props} />
      <Accordion
        id="accordion-4"
        togglePosition="left"
        subtitle="Optional Subtitle"
        useDivider
        {...props}
      />
    </div>
  );
}
