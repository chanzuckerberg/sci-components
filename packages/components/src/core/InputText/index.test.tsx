import { generateSnapshots } from "@chanzuckerberg/story-utils";
import { composeStories } from "@storybook/react";
import { render, screen } from "@testing-library/react";
import * as stories from "./index.stories";

// Returns a component that already contain all decorators from story level, meta level and global level.
const { Test } = composeStories(stories);

describe("<InputText />", () => {
  generateSnapshots(stories);

  it("renders inputText component", () => {
    render(<Test {...Test.args} />);
    const inputTextElement = screen.getByTestId("inputTextBase");
    expect(inputTextElement).not.toBeNull();
  });

  it("component fails if missing id or label prop", () => {
    render(<Test />);
    const inputTextElement = screen.queryByTestId("inputTextFail");
    expect(inputTextElement).toBeNull();
  });

  it("hides label is hideLabel is true, but aria-label is still set", () => {
    render(<Test />);
    const inputTextElement = screen.getByLabelText("Hidden Label");
    expect(inputTextElement).not.toBeNull();
  });

  it("renders text area if sdsType is set to 'textArea'", () => {
    render(<Test />);
    const inputTextAreaElement = screen.getByTestId("inputTextArea");
    expect(inputTextAreaElement).not.toBeNull();
  });
});
