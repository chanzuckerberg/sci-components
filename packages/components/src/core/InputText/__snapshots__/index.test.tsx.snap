// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputText /> Default story renders snapshot 1`] = `
<label
  class="css-1n9rsuo"
  for="Test"
>
  Label
</label>
`;
